File: Most Profit Assigning Work.cpp
class Solution {
public:
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
        vector<pair<int, int>> q;
        int n = profit.size();
        int min_difficulty = 1 << 17;
        int max_profit = 0;
        for (int i = 0; i < n; i++) {
            min_difficulty = min(min_difficulty, difficulty[i]);
            max_profit = max(max_profit, profit[i]);
            q.emplace_back(difficulty[i], -profit[i]);
        }

        sort(q.begin(), q.end());
        int maxx = 0;
        for (int i = 0; i < n; i++) {
            maxx = min(maxx, q[i].second);
            q[i].second = maxx;
        }

        int total = 0;
        for (int w : worker) {
            if (w < min_difficulty) continue;
            auto lower = lower_bound(q.begin(), q.end(), make_pair(w, -max_profit));
            if (lower != q.end() && lower->first == w) total -= lower->second;
            else total -= (lower-1)->second;
        }
        return total;
    }
};

File: Most Profit Assigning Work.go
type Pair struct {
    difficulty int
    profit int
}

func maxProfitAssignment(difficulty []int, profit []int, worker []int) int {
    q := make([]Pair, 0)
    n := len(profit)
    min_difficulty := 1 << 17
    max_profit := 0
    for i := 0; i < n; i++ {
        min_difficulty = min(min_difficulty, difficulty[i])
        max_profit = max(max_profit, profit[i])
        q = append(q, Pair{difficulty: difficulty[i], profit: -profit[i]})
    }

    sort.Slice(q, func(i, j int) bool {
                        if q[i].difficulty == q[j].difficulty {
                            return q[i].profit < q[j].profit
                        }
                        return q[i].difficulty < q[j].difficulty
                    })

    maxx := 0
    for i := 0; i < n; i++ {
        maxx = min(maxx, q[i].profit)
        q[i].profit = maxx
    }

    total := 0
    for _, w := range worker {
        if w < min_difficulty {
            continue
        }

        i := sort.Search(len(q), func(i int) bool { return q[i].difficulty >= w })
        if i < n && w == q[i].difficulty {
            total -= q[i].profit
        } else {
            total -= q[i-1].profit
        }
    }
    return total
}

